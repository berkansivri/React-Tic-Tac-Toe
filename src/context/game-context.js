import React, { useContext } from 'react'

const gameContext = React.createContext()

const useGameContext = () => useContext(gameContext)

export { gameContext as default, useGameContext }
