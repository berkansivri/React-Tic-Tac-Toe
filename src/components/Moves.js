import React from 'react'
import { useGameContext } from '../context/game-context' 
import { setNewGame } from '../hooks/useCheckWinner'

const Moves = () => {
  const { moves, dispatch, setTurn } = useGameContext()
  
  const handleGoMove = (step) => {
    dispatch({type: "GO_MOVE", step })
    setTurn(moves[step].sign === "X" ? "O" : "X")
  }

  const handleNewGame = () => {
    setNewGame(dispatch)
  }

  const getMoves = () => {
    return moves.map(({ location, sign }, index) => {
      return (
        <li key={location} style={{ height: '35px' }}>
          <p style={{ display: "inline-block" }}>
            {`${sign} to position [${Math.floor(location / 3 + 1)},${location % 3 + 1}]`}
          </p>
          <button style={{ marginLeft: '5px' }} id="btnGoMove" onClick={() => handleGoMove(index)}>
            Go
          </button>
        </li>
      )
    })
  }

  return (
    <>
      <button id="btnNewGame" style={{ marginLeft: '20px' }} onClick={handleNewGame}>
        New Game
      </button>
      <ol style={{ marginTop: '0px' }}>
        {getMoves()}
      </ol>
    </>
  )
}

export { Moves as default }