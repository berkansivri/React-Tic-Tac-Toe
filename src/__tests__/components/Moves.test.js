import React from 'react'
import { shallow } from 'enzyme'
import Moves from '../../components/Moves'
import * as Context from '../../context/game-context'
import moves from '../fixtures/moves'

let contextValues, wrapper

beforeEach(() => {
  contextValues = {
    dispatch: jest.fn(),
    setTurn: jest.fn(),
    moves
  }

  jest.spyOn(Context, 'useGameContext')
      .mockImplementation(() => contextValues)
  
  wrapper = shallow(<Moves />)
})

test("should render Moves component correctly", () => {
  expect(wrapper).toMatchSnapshot()
})

test("should render all moves correctly", () => {
  expect(wrapper.find("li")).toHaveLength(moves.length)
})

test("should handle on click new game", () => {
  wrapper.find("#btnNewGame").simulate("click")
  expect(contextValues.dispatch).toHaveBeenCalledWith({ type:"CLEAR_MOVES" })
})

test("should handle go move", () => {
  wrapper.find("#btnGoMove").first().simulate("click")
  expect(contextValues.dispatch).toHaveBeenCalledWith({type: "GO_MOVE", step: 0 })
  expect(contextValues.setTurn).toHaveBeenCalledWith(moves[0].sign === "X" ? "O" : "X")
})