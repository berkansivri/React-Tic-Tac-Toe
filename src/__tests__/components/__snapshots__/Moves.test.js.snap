// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Moves component correctly 1`] = `
<Fragment>
  <button
    id="btnNewGame"
    onClick={[Function]}
    style={
      Object {
        "marginLeft": "20px",
      }
    }
  >
    New Game
  </button>
  <ol
    style={
      Object {
        "marginTop": "0px",
      }
    }
  >
    <li
      key="0"
      style={
        Object {
          "height": "35px",
        }
      }
    >
      <p
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        X to position [1,1]
      </p>
       
      <button
        id="btnGoMove"
        onClick={[Function]}
      >
        Go
      </button>
    </li>
    <li
      key="1"
      style={
        Object {
          "height": "35px",
        }
      }
    >
      <p
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        O to position [1,2]
      </p>
       
      <button
        id="btnGoMove"
        onClick={[Function]}
      >
        Go
      </button>
    </li>
    <li
      key="2"
      style={
        Object {
          "height": "35px",
        }
      }
    >
      <p
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        X to position [1,3]
      </p>
       
      <button
        id="btnGoMove"
        onClick={[Function]}
      >
        Go
      </button>
    </li>
    <li
      key="3"
      style={
        Object {
          "height": "35px",
        }
      }
    >
      <p
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        O to position [2,1]
      </p>
       
      <button
        id="btnGoMove"
        onClick={[Function]}
      >
        Go
      </button>
    </li>
    <li
      key="4"
      style={
        Object {
          "height": "35px",
        }
      }
    >
      <p
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        X to position [2,2]
      </p>
       
      <button
        id="btnGoMove"
        onClick={[Function]}
      >
        Go
      </button>
    </li>
    <li
      key="8"
      style={
        Object {
          "height": "35px",
        }
      }
    >
      <p
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        X to position [3,3]
      </p>
       
      <button
        id="btnGoMove"
        onClick={[Function]}
      >
        Go
      </button>
    </li>
  </ol>
</Fragment>
`;
